File: zigzagConversion/zigzagConversion.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    string convert(string s, int numRows) {
        vector<string> rows(numRows);
        int i = 0;
        while (i < s.length()) {
            for (int j = 0; j < numRows && i < s.length(); j++) { // up to down
                rows[j] += s[i++];
            }
            for (int j = numRows - 2; j >= 1 && i < s.length(); j--) {
                rows[j] += s[i++];
            }
        }
        string res = "";
        for (string& row : rows)
            res += row;
        return res;
    }
};

int main()
{
  Solution s;
  string ss = "ABCDE";
  cout << s.convert(ss, 3) << endl;
  return 0;
}

File: 6.zigzagConversion/zigzagConversion.java
class Solution {
    public String convert(String s, int numRows) {
        StringBuilder[] sb = new StringBuilder[numRows];
        for (int i = 0; i < numRows; i++) sb[i] = new StringBuilder();
        char[] arr = s.toCharArray();
        int i = 0;
        while (i < arr.length) {
            for (int j = 0; j < numRows && i < arr.length; j++)
                sb[j].append(arr[i++]);
            for (int j = numRows - 2; j >= 1 && i < arr.length; j--)
                sb[j].append(arr[i++]);
        }
        for (i = 1; i < sb.length; i++)
            sb[0].append(sb[i]);
        return sb[0].toString();
    }
}
